packages/core/src/config.ts

```diff
@@ -88,6 +88,108 @@ export async function loadConfig({
   return content as RslibConfig;
 }
 
+const composeExternalsWarnConfig = (
+  format: Format,
+  ...externalsArray: NonNullable<RsbuildConfig['output']>['externals'][]
+): RsbuildConfig => {
+  if (format !== 'esm') {
+    return {};
+  }
+
+  const externals: NonNullable<RsbuildConfig['output']>['externals'] = [];
+  for (const e of externalsArray.filter(Boolean)) {
+    if (Array.isArray(e)) {
+      externals.push(...e);
+    } else {
+      // @ts-ignore
+      externals.push(e);
+    }
+  }
+
+  // Match logic is derived from https://github.com/webpack/webpack/blob/94aba382eccf3de1004d235045d4462918dfdbb7/lib/ExternalModuleFactoryPlugin.js#L166-L293.
+  const matchUserExternals = (
+    externals: NonNullable<RsbuildConfig['output']>['externals'],
+    request: string,
+    callback: (matched?: true) => void,
+  ) => {
+    if (typeof externals === 'string') {
+      if (externals === request) {
+        callback(true);
+        return;
+      }
+    } else if (Array.isArray(externals)) {
+      let i = 0;
+      const next = () => {
+        let asyncFlag: boolean;
+        const handleExternalsAndCallback = (matched?: true) => {
+          if (!matched) {
+            if (asyncFlag) {
+              asyncFlag = false;
+              return;
+            }
+            return next();
+          }
+
+          callback(matched);
+        };
+
+        do {
+          asyncFlag = true;
+          if (i >= externals.length) {
+            return callback();
+          }
+          matchUserExternals(
+            externals[i++],
+            request,
+            handleExternalsAndCallback,
+          );
+        } while (!asyncFlag);
+        asyncFlag = false;
+      };
+
+      next();
+      return;
+    } else if (externals instanceof RegExp) {
+      if (externals.test(request)) {
+        callback(true);
+      }
+      return;
+    } else if (typeof externals === 'function') {
+      // TODO: Support function
+    } else if (typeof externals === 'object') {
+      if (Object.prototype.hasOwnProperty.call(externals, request)) {
+        callback(true);
+        return;
+      }
+    }
+
+    callback();
+  };
+
+  return {
+    output: {
+      externals: [
+        ({ request, dependencyType, contextInfo }: any, callback: any) => {
+          let externalized = false;
+          const _callback = (matched?: true) => {
+            if (matched) {
+              externalized = true;
+            }
+          };
+
+          if (contextInfo.issuer && dependencyType === 'commonjs') {
+            matchUserExternals(externals, request, _callback);
```

Comment:
Why not return boolean directly insteadof using callbacl function?

Reply:
It's somehow a complex logic, just to port webpack's logic and edit as less as possible.